sdk/core/System.Net.ClientModel/src/Message/ClientMessage.cs

```diff
@@ -0,0 +1,102 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Net.ClientModel.Internal;
+using System.Threading;
+
+namespace System.Net.ClientModel.Core;
+
+public class ClientMessage : IDisposable
+{
+    private MessageResponse? _response;
+    private bool _disposed;
+
+    protected internal ClientMessage(MessageRequest request)
+    {
+        Request = request;
+        _propertyBag = new ArrayBackedPropertyBag<ulong, object>();
+    }
+
+    public virtual MessageRequest Request { get; }
+
+    public virtual MessageResponse Response
+    {
+        get
+        {
+            if (_response is null)
+            {
+                throw new InvalidOperationException("Response has not been set on Message.");
+            }
+
+            return _response;
+        }
+
+        // This is set internally by the transport.
+        protected internal set => _response = value;
+    }
+
+    public bool HasResponse => _response is not null;
+
+    #region Pipeline invocation options
+
+    public virtual CancellationToken CancellationToken { get; set; }
+
+    private ArrayBackedPropertyBag<ulong, object> _propertyBag;
+
+    public bool TryGetProperty(Type type, out object? value) =>
+        _propertyBag.TryGetValue((ulong)type.TypeHandle.Value, out value);
+
+    public void SetProperty(Type type, object value) =>
+        _propertyBag.Set((ulong)type.TypeHandle.Value, value);
+
+    private MessageClassifier? _messageClassifer;
+    public virtual MessageClassifier MessageClassifier
+    {
+        get
+        {
+            if (_messageClassifer is null)
+            {
+                throw new InvalidOperationException("MessageClassifer cannot be accessed before it is set on the message.");
```

Comment:
why not make the property nullable so that callers can check if it is set without having to catch?